scripts: build: elf_parser: fix relocatable data offsets

JordanYates opened the pull request:
Retrieve the section holding the symbol data directly from the symbol information, instead of trying to find a section that contains data at the correct address. The later approach does not work for relocatable files, which contain multiple sections all containing data in overlapping temporary memory addresses (usually starting at 0).

Based on investigations with relocatable `.elf` files, the symbol table data in relocatable files is not shifted by the section address. Remove the shift from these files to ensure that data can be pulled from the symbol table.

Problems were investigated at the request of @aescolar, and correct parsing was validated with a relocatable `.elf` file that he provided. Tests will validate that this change does not break standard applications, but will not test the relocatable case (which is not currently tested anyway).
```
jordan@FIDO:~/code/zephyr$ ./zephyr/scripts/build/gen_handles.py -o handles.c -s __device_start -z zephyr -k zephyr_pre0.reloc.elf
jordan@FIDO:~/code/zephyr$ ./zephyr/scripts/build/gen_handles.py -o handles.c -s __device_start -z zephyr -k zephyr_pre0.normal.elf
```

Commit:
scripts: build: elf_parser: retrieve section directly

Retrieve the section holding the symbol data directly from the symbol
information, instead of trying to find a section that contains data at
the correct address. The later approach does not work for relocatable
files, which contain multiple sections all containing data in
overlapping temporary memory addresses (usually starting at 0).

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
scripts: build: elf_parser: fix relocatable data offsets

Based on investigations with relocatable `.elf` files, the symbol table
data in relocatable files is not shifted by the section address. Remove
the shift from these files to ensure that data can be pulled from the
symbol table.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

aescolar reviewed (dismissed):
Thanks @JordanYates .
The change is welcome on its own: Even that with it alone we cannot yet get gen_handles.py to work fully with relocatable elf files, given that elf_parser is a common module, and this makes it more general, and faster, I think we should merge it.

As a possible follow up (maybe another PR), if gen_handles does not get to work with relocatable elf's, maybe we should have gen_handles check if the elf was relocatable or executable, and if relocatable print a clear error like "gen_handles does not support yet relocatable elf files" (for whoever may come after). What do you think?

Commit:
scripts: build: gen_handles: exit on relocatable files

Print a clean error message when provided a relocatable file, instead
of generating a source file which will cause compilation errors further
away from the source.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

aescolar commented on scripts/build/gen_handles.py:
```diff
@@ -121,6 +121,13 @@ def main():
+        # symbol table are invalid (as expected, because the structures have not
+        # yet been allocated addresses). Fixing this will require iterating over
+        # the relocation sections to find the symbols those pointers will end up
+        # referring too.
```
too -> to
